### PR TITLE
split: pass GNU tests/b-chunk.sh

### PR DESCRIPTION
**Pull request by zhitkoff**
This PR adds 
* full implementation for `---io-blksize` option
* refactoring of how splitting into N chunks works for `--number` strategies
* handling input size of stdin stream and file size for files at /dev, /proc, /sys and similar locations that either report `0` for file size, while actually having some content OR report size greater than actual content

passes GNU tests/b-chunk.sh

### CODE REVIEW

**Comment by zhitkoff** [comment]
@tertsdiepraam would you mind reviewing this one ?

**Comment by zhitkoff** [comment]
setting back to draft - even though it passes the GNU test, the implementation for ---io-blksize option is incomplete, needs more work

**Comment by zhitkoff** [comment]
@tertsdiepraam @sylvestre this one is ready for review

File: src/uu/split/src/split.rs [hunk 1]
```diff
@@ -499,12 +517,29 @@ impl Settings {
...
                     "\\0" => b'\0',
                     s if s.as_bytes().len() == 1 => s.as_bytes()[0],
-                    s => return Err(SettingsError::MultiCharacterSeparator(s.to_owned())),
+                    s => return Err(SettingsError::MultiCharacterSeparator(s.to_string())),
                 }
             }
             None => b'\n',
         };
 
+        let io_blksize: Option<usize> = if let Some(s) = matches.get_one::<String>(OPT_IO_BLKSIZE) {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:526]: What do you think about ?
let io_blksize: Option<usize> = matches.get_one::<String>(OPT_IO_BLKSIZE).map(|s| {
    parse_size_u64(s)
        .ok()
        .and_then(|n| {
            let n = n.try_into().map_err(|_| SettingsError::InvalidIOBlockSize(s.to_string()))?;
            if n > OPT_IO_BLKSIZE_MAX {
                Err(SettingsError::InvalidIOBlockSize(s.to_string()))
            } else {
                Ok(n)
            }
        })
}).transpose()?;

File: src/uu/split/src/split.rs [hunk 2]
```diff
@@ -591,6 +627,134 @@ fn custom_write_all<T: Write>(
...
+            format!("{}: cannot determine input size", input),
+        ));
+    } else {
+        // Could be that file size is larger than set limit
+        // or could be an "infinite" file like /dev/zero
+        // Attempt direct `seek()` for the end of a file
+        let mut tmp_fd = File::open(Path::new(input))?;
+        let end = tmp_fd.seek(SeekFrom::End(0))?;
+        if end > 0 {
+            input_size = end;
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:694]: same:
```suggestion
            Ok(end)
```

File: src/uu/split/src/split.rs [hunk 3]
```diff
@@ -591,6 +627,134 @@ fn custom_write_all<T: Write>(
...
+
+    // Set read limit to io_blksize if specified
+    // Otherwise to OPT_IO_BLKSIZE_MAX
+    let read_limit: u64 = if let Some(n) = io_blksize {
+        *n
+    } else {
+        OPT_IO_BLKSIZE_MAX
+    }
+    .try_into()
+    .unwrap();
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:663]:
```suggestion
    let read_limit = io_blksize.unwrap_or(OPT_IO_BLKSIZE_MAX) as u64;
```

File: src/uu/split/src/split.rs [hunk 4]
```diff
@@ -591,6 +627,134 @@ fn custom_write_all<T: Write>(
...
+        || input.starts_with("/dev/")
+        || input.starts_with("/proc/")
+        || input.starts_with("/sys/")
+    {
+        // STDIN or
+        // Files in system Unix and Linux filesystems which
+        // could report incorrect file size via `metadata.len()`
+        // either `0` while there is content
+        // or a size larger than actual content
+        input_size = get_irregular_input_size(input, reader, buf, io_blksize)?;
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:745]: same, early return

```suggestion
        get_irregular_input_size(input, reader, buf, io_blksize)
```

File: src/uu/split/src/split.rs [hunk 5]
```diff
@@ -591,6 +627,134 @@ fn custom_write_all<T: Write>(
...
+    } else {
+        // Regular file
+        let metadata = metadata(input)?;
+        input_size = metadata.len();
+        // Double check the size if `metadata.len()` reports `0`
+        if input_size == 0 {
+            input_size = get_irregular_input_size(input, reader, buf, io_blksize)?;
+        }
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:753]:
```suggestion       // Regular file
        let metadata = metadata(input)?;
        let size = metadata.len();
        // If metadata reports size 0, use get_irregular_input_size to double-check
        if size == 0 {
            get_irregular_input_size(input, reader, buf, io_blksize)
        } else {
            Ok(size)
        }
```

File: src/uu/split/src/split.rs (hunk 1 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:526]: in this case it would "eat" possible error returned by parse_size_u64(), so it cannot be bubbled up to from()->Result. Also, since ok() converts it to Option and the and_then() expects an option to be returned from the closure, it is not straitforward to return the SettingsError in either places within the closure inside closure - i.e. doing `return Err(SettingsError::InvalidIOBlockSize(s.to_string()))` or `Some(Err(SettingsError::InvalidIOBlockSize(s.to_string())))` would not work : as return is not for from()->Result function but closure, and wrapping in Some() to satisfy and_then() signature we will end up with `Option<Option<Result<>>>` as a retuning value from closure from map() instead of `Option<Result<>>` ... could unwrap(), but then still missing/eating error from parse_size_u64() ... 
I could be missing something though

Comment (sylvestre) [reply src/uu/split/src/split.rs:526]: please ignore my comment then :(

File: src/uu/split/src/split.rs [hunk 6]
```diff
@@ -40,11 +41,23 @@ static OPT_HEX_SUFFIXES_SHORT: &str = "-x";
...
+// Could be due to how `std::io::BufReader` handles internal buffers.
+// So we use much smaller value for those
+static OPT_IO_BLKSIZE_MAX: usize = if usize::BITS >= 64 {
+    2_146_435_072
+} else {
+    1_000_000_000
+};
+// The ---io parameter is consumed and ignored.
+// The parameter is included to make GNU coreutils tests pass.
+static OPT_IO: &str = "-io";
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:60]: Here's an example from GNU:
```
❯ split ---
split: option '---io-blksize' requires an argument
Try 'split --help' for more information.
❯ split ---io-bl
split: option '---io-blksize' requires an argument
Try 'split --help' for more information.
```

This is not a separate option, but one of those things where a long argument can be shortened to the first few letters if its unambiguous. `clap` supports this too and I think it's turned on (it's `infer_long_args(true)`) so do we need this?

File: src/uu/split/src/split.rs [hunk 7]
```diff
@@ -1018,155 +1171,118 @@ impl<'a> Write for LineBytesChunkWriter<'a> {
...
 ) -> UResult<()>
 where
-    R: Read,
+    R: BufRead,
 {
-    // Get the size of the input file in bytes and compute the number
-    // of bytes per chunk.
-    //
+    // Get the size of the input in bytes
+    let initial_buf: &mut Vec<u8> = &mut vec![];
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:1213]:
```suggestion
    let initial_buf = Vec::new();
```

File: src/uu/split/src/split.rs [hunk 8]
```diff
@@ -1018,155 +1171,118 @@ impl<'a> Write for LineBytesChunkWriter<'a> {
...
-    for _ in 0..num_chunks {
-        let filename = filename_iterator
-            .next()
-            .ok_or_else(|| USimpleError::new(1, "output file suffixes exhausted"))?;
-        let writer = settings.instantiate_current_writer(filename.as_str())?;
-        writers.push(writer);
+    if usize::BITS < 64 {
+        let _: usize = num_chunks
+            .try_into()
+            .map_err(|_| USimpleError::new(1, "Number of chunks too big"))?;
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:1251]: Why this change? The original code makes sense to me. If we need it to fit in a `usize` why wouldn't we make it a `usize`?

File: src/uu/split/src/split.rs [hunk 9]
```diff
@@ -591,6 +627,123 @@ fn custom_write_all<T: Write>(
...
+    buf: &mut Vec<u8>,
+    io_blksize: &Option<usize>,
+) -> std::io::Result<u64>
+where
+    R: BufRead,
+{
+    if input == "-"
+        || input.starts_with("/dev/")
+        || input.starts_with("/proc/")
+        || input.starts_with("/sys/")
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:726]: This feels inherently brittle and I'm wondering whether there's a better way to do it. For example, we could have links to all of these filesystems and that should still work. Based on the syscalls they make, I don't think GNU has exceptions like, though I'm not entirely sure what they are doing instead.

Comment (tertsdiepraam) [reply src/uu/split/src/split.rs:726]: I think they might just try to get the length first and try this other method if that returns `0`? I'm not sure.

File: src/uu/split/src/split.rs [hunk 10]
```diff
@@ -1018,155 +1171,118 @@ impl<'a> Write for LineBytesChunkWriter<'a> {
...
-
-    let chunk_size = (num_bytes / (num_chunks)).max(1);
-    let mut num_bytes: usize = num_bytes.try_into().unwrap();
-
-    let mut i = 1;
+    let mut i = 0;
     loop {
+        let chunk_size = chunk_size_base + (chunk_size_reminder > i) as u64;
         let buf: &mut Vec<u8> = &mut vec![];
+        i += 1;
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:1285]: Would this work?
```rust
for i in 1.. {
    ...
}
```

**Review by tertsdiepraam — COMMENTED** [review]
Impressive work! Seeing that GNU test pass is great!

File: src/uu/split/src/split.rs (hunk 6 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:60]: yep, looks like it is the case. will remove

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:726]: in GNU they first go straight to reading it into a buffer and then compare that to what metadata/filesystem is reporting + do seek on the file if those disagree (as well as trying to copy things arround into temp file/buffer, etc) and some other edge cases. I tried to avoid reading file into a buffer first for everything, but could re-implement to closer follow GNU approach

File: src/uu/split/src/split.rs (hunk 8 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:1251]: it made sense for the original code because num_chunks was later used in method calls that require usize like `writers.iter_mut().take(num_chunks - 1)`, and there was a test to test for that error - looks like only for code coverage %. I left it in to keep the test, but looking at it more closely - it does not need to be usize, so I will remove it along with the test that checked for it.

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (tertsdiepraam) [reply src/uu/split/src/split.rs:726]: I think doing what GNU does is the right call! Nice investigative work! (As a precaution: remember not to look at the GNU source code)

**Comment by zhitkoff** [comment]
@tertsdiepraam @sylvestre 
made changes based on review, please let me know if you have any more suggestions/comments

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs]: should be ready to go

**Comment by zhitkoff** [comment]
@tertsdiepraam @sylvestre I have another set of changes ready that pass next Gnu test for split  'l-chunk', but they are based on code included in this PR. Should I add them into this PR or wait until this one is resolved and then submit those in a new one?
